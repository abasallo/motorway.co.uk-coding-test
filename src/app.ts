import * as dotenv from 'dotenv'

import express from 'express'
import cors from 'cors'
import bodyParser from 'body-parser'

import index from './express/routes'
import vehicle from './express/routes/vehicle'

import { initialiseSequelize } from './orm/initialiseSequelize'
import { initialiseTestDatabase } from './utils/initialiseTestDatabase'
import { errorMiddleware } from './express/errors'
import { AppOrmModel } from './orm/model/app.orm.model'
dotenv.config()

export const app = express()

export const model: Promise<AppOrmModel> = process.env.CONNECTION_URL
  ? initialiseSequelize(process.env.CONNECTION_URL)
  : initialiseTestDatabase()

app.use(cors())

app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.use('/', index)
app.use('/vehicle', vehicle)

app.use(errorMiddleware)

export default app
