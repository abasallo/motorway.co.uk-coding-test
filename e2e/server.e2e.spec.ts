import axios from 'axios'

import { ReasonPhrases, StatusCodes } from 'http-status-codes'

import app from '../src/app'
import http from 'http'

const TEST_SERVER_PORT = 4001

const ids = [1, 1, 1, 2, 2, 2, 3, 3, 3, 3]
const timestamps = [
  '2022-09-01 00:00:00+00',
  '2022-09-10 10:23:54+00',
  '2022-09-11 00:00:00+00',
  '2022-09-11 16:00:00+00',
  '2022-09-11 17:03:17+00',
  '2022-09-11 18:00:00+00',
  '2022-09-12 11:00:00+00',
  '2022-09-12 12:41:41+00',
  '2022-09-13 00:00:00+00',
  'wrong'
]

const expectedStatusCodes = [
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK,
  StatusCodes.OK
]

const expectedStatusTexts = [
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK,
  ReasonPhrases.OK
]

let server: http.Server
beforeAll(async () => {
  server = app.listen({ port: TEST_SERVER_PORT }, () => console.log(`Test Server initialised on port: ${TEST_SERVER_PORT}`))
})

afterAll(() => {
  server.close()
})

describe('Server E2E tests', () => {
  describe('Vehicle E2E tests', () => {
    for (const [index, id] of ids.entries()) {
      test(`Expected Vehicle Info for id: ${id} and timestamp: ${timestamps[index]}`, async () => {
        const { status, statusText, data } = await axios({
          method: 'get',
          url: `http://localhost:${TEST_SERVER_PORT}/vehicle/`,
          data: {
            id,
            timestamp: timestamps[index]
          }
        })

        expect(status).toEqual(expectedStatusCodes[index])
        expect(statusText).toEqual(expectedStatusTexts[index])

        expect(data).toMatchSnapshot()
      })
    }

    test('Non existent ID', async () => {
      try {
        await axios({
          method: 'get',
          url: `http://localhost:${TEST_SERVER_PORT}/vehicle/`,
          data: {
            id: -1,
            timestamp: timestamps[0]
          }
        })
      } catch (error) {
        expect(error).toBeDefined()
      }
    })

    test('Wrong format ID', async () => {
      try {
        await axios({
          method: 'get',
          url: `http://localhost:${TEST_SERVER_PORT}/vehicle/`,
          data: {
            id: 'wrong',
            timestamp: timestamps[0]
          }
        })
      } catch (error) {
        expect(error).toBeDefined()
      }
    })
  })
  describe('Health Check E2E tests', () => {
    test('Expected response for Health Check', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })
})
