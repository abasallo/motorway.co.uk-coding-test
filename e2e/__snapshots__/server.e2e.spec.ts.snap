// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server E2E tests Health Check E2E tests Expected response for Health Check 1`] = `"OK"`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 1 and timestamp: 2022-09-01 00:00:00+00 1`] = `
{
  "timestamp": "2022-09-01 00:00:00+00",
  "vehicle": {
    "id": 1,
    "make": "BMW",
    "model": "X1",
    "state": "quoted",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 1 and timestamp: 2022-09-10 10:23:54+00 1`] = `
{
  "state": "quoted",
  "timestamp": "2022-09-10 10:23:54+00",
  "vehicle": {
    "id": 1,
    "make": "BMW",
    "model": "X1",
    "state": "quoted",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 1 and timestamp: 2022-09-11 00:00:00+00 1`] = `
{
  "state": "quoted",
  "timestamp": "2022-09-11 00:00:00+00",
  "vehicle": {
    "id": 1,
    "make": "BMW",
    "model": "X1",
    "state": "quoted",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 2 and timestamp: 2022-09-11 16:00:00+00 1`] = `
{
  "state": "quoted",
  "timestamp": "2022-09-11 16:00:00+00",
  "vehicle": {
    "id": 2,
    "make": "AUDI",
    "model": "A4",
    "state": "selling",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 2 and timestamp: 2022-09-11 17:03:17+00 1`] = `
{
  "state": "selling",
  "timestamp": "2022-09-11 17:03:17+00",
  "vehicle": {
    "id": 2,
    "make": "AUDI",
    "model": "A4",
    "state": "selling",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 2 and timestamp: 2022-09-11 18:00:00+00 1`] = `
{
  "state": "selling",
  "timestamp": "2022-09-11 18:00:00+00",
  "vehicle": {
    "id": 2,
    "make": "AUDI",
    "model": "A4",
    "state": "selling",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 3 and timestamp: 2022-09-12 11:00:00+00 1`] = `
{
  "state": "selling",
  "timestamp": "2022-09-12 11:00:00+00",
  "vehicle": {
    "id": 3,
    "make": "VW",
    "model": "GOLF",
    "state": "sold",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 3 and timestamp: 2022-09-12 12:41:41+00 1`] = `
{
  "state": "sold",
  "timestamp": "2022-09-12 12:41:41+00",
  "vehicle": {
    "id": 3,
    "make": "VW",
    "model": "GOLF",
    "state": "sold",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 3 and timestamp: 2022-09-13 00:00:00+00 1`] = `
{
  "state": "sold",
  "timestamp": "2022-09-13 00:00:00+00",
  "vehicle": {
    "id": 3,
    "make": "VW",
    "model": "GOLF",
    "state": "sold",
  },
}
`;

exports[`Server E2E tests Vehicle E2E tests Expected Vehicle Info for id: 3 and timestamp: wrong 1`] = `
{
  "timestamp": "wrong",
  "vehicle": {
    "id": 3,
    "make": "VW",
    "model": "GOLF",
    "state": "sold",
  },
}
`;
